### DISCUSSION TITLE
时间戳的插入与创建 (How to create and insert Timestamp/Time)

### DISCUSSION THREAD

**Discussion by male168** [apache/iotdb#3736] [Q&A] [answered]
（Python）请问如何插入和创建时间戳，并且时间戳 的类型和格式应该是什么样的，我创建了一些时间序列并插入了一些数据，读取时有一列Time,自动生成一些id。

**Comment by qiaojialin** [comment, marked as answer]
Hi, each unary timeseries has two columns (Timestamp/Time, Value). The data type of Timestamp/Time is long/INT64. 

Timestamp/Time is a build-in column in timeseries. It can not be created explicitly, you can only create timeseries and insert the timestamp.

E.g., ```insert into root.turbine1.d1(timestamp,s1) values(1,2)```
This will insert a timestamp=1 and value=2 into timeseries root.turbine1.d1.s1.

When doing queries, the Timestamp/Time will be printed automatically.

The id is created by the client, not IoTDB.

**Reply by male168** [reply]
Should we create a time column in INT64 type? 那我们在python中插入时间戳，应该自己创建一条时间列，然后类型为INT64吗

E.g.：
measurements_ = ["Time","s_01", "s_02", "s_03", "s_04", "s_05", "s_06"]
values_ = [1628829206860,False, 20, 20, 1.1, 11.1, "test_record2"]
data_types_ = [
    TSDataType.INT64,
    TSDataType.BOOLEAN,
    TSDataType.INT32,
    TSDataType.INT64,
    TSDataType.FLOAT,
    TSDataType.DOUBLE,
    TSDataType.TEXT,]
session.insert_record("root.sg_test_01.d_01", 1, measurements_, data_types_, values_)

**Reply by qiaojialin** [reply]
No need to create Timestamp/Time column.
When inserting, multiple measurements could share one time.

In this example:
measurements_ = ["s_01", "s_02", "s_03", "s_04", "s_05", "s_06"]
values_ = [False, 20, 20, 1.1, 11.1, "test_record2"]
data_types_ = [
TSDataType.BOOLEAN,
TSDataType.INT32,
TSDataType.INT64,
TSDataType.FLOAT,
TSDataType.DOUBLE,
TSDataType.TEXT,]
session.insert_record("root.sg_test_01.d_01", 1, measurements_, data_types_, values_)

Time is not a part of the values.

**Reply by male168** [reply]
我用客户端查询时time列显示的1970年，为什么通过上述例子插入数据时生成的不是最新的时间戳呢

**Reply by qiaojialin** [reply]
时间戳的 0 是从 1970 年开始计时的。insert_record 时候 time 参数写了个 1，1就是 1970年1月1日0点0分0秒001毫秒。
可以 用 System.currentTimeMillis() 获取当前时间戳，不过注意这个调用频繁可能有重复.

Timestamp is timed since 1970-1-1 00:00:00.000
You could use  System.currentTimeMillis() to get current timestamp.

Or use now() in sql: ```insert into root.sg.d1(time, s1) values(now(), 1)```

**Reply by male168** [reply]
解决了，感谢

**Reply by jixuan1989** [reply]
> 我用客户端查询时time列显示的1970年，为什么通过上述例子插入数据时生成的不是最新的时间戳呢

see your codes:

> session.insert_record("root.sg_test_01.d_01", 1, measurements_, data_types_, values_)

the second field, 1, is the timestamp.

Timestamp means how many milliseconds the time has elapsed from 1970.01.01, it is a default behavior in many program language, e.g. Java, Python.

You can change your codes like this:

```python

import time

millis = int(round(time.time() * 1000))
session.insert_record("root.sg_test_01.d_01", millis, measurements_, data_types_, values_)
 ```

Reference:
https://www.runoob.com/python3/python3-date-time.html

**Comment by cigarl** [comment]
时间戳的类型和格式，这里有详细说明：https://iotdb.apache.org/zh/UserGuide/Master/Data-Concept/Data-Type.html
